src/main/java/software/aws/neptune/NeptuneDriver.java

```diff
@@ -76,62 +68,75 @@ public boolean acceptsURL(final String url) throws SQLException {
-            throws SQLException {
-        // TODO AN-550: Switch to map with class that holds Conn properties, key, query executor, etc.
-        if (OPENCYPHER.equalsIgnoreCase(language)) {
-            return new OpenCypherConnectionProperties(properties);
-        }
-        if (GREMLIN.equalsIgnoreCase(language) || SQL_GREMLIN.equalsIgnoreCase(language)) {
-            return new GremlinConnectionProperties(properties);
-        }
-        if (SPARQL.equalsIgnoreCase(language)) {
-            return new SparqlConnectionProperties(properties);
-        }
-        return new OpenCypherConnectionProperties(properties);
-    }
+    private class DriverConnectionFactory {
+        private final Pattern connStringPattern = Pattern.compile(CONN_STRING_PREFIX + "(\\w+)://(.*)");
+        private static final String OPENCYPHER = "opencypher";
+        private static final String GREMLIN = "gremlin";
+        private static final String SQL_GREMLIN = "sqlgremlin";
+        private static final String SPARQL = "sparql";
 
-    private String firstPropertyKey(final String language) {
-        // TODO AN-550: Switch to map with class that holds Conn properties, key, query executor, etc.
-        if (OPENCYPHER.equalsIgnoreCase(language)) {
-            return OpenCypherConnectionProperties.ENDPOINT_KEY;
+        public java.sql.Connection getConnection(final String url, final Properties info) throws SQLException {
+            final String language = getLanguage(url, connStringPattern);
+            final String propertyString = getPropertyString(url, connStringPattern);
+            final String firstPropertyKey = getFirstPropertyKey(language);
+            final Properties properties = setProperties(propertyString, firstPropertyKey, info);
+
+            if (OPENCYPHER.equalsIgnoreCase(language)) {
+                return new OpenCypherConnection(new OpenCypherConnectionProperties(properties));
+            }
+            if (GREMLIN.equalsIgnoreCase(language)) {
+                return new GremlinConnection(new GremlinConnectionProperties(properties));
+            }
+            if (SQL_GREMLIN.equalsIgnoreCase(language)) {
+                return new SqlGremlinConnection(new GremlinConnectionProperties(properties));
+            }
+            if (SPARQL.equalsIgnoreCase(language)) {
+                return new SparqlConnection(new SparqlConnectionProperties(properties));
+            }
+            LOGGER.error("Encountered Neptune JDBC connection string but failed to parse driver language.");
+            throw SqlError.createSQLException(
+                    LOGGER, SqlState.CONNECTION_EXCEPTION,
+                    SqlError.INVALID_CONNECTION_PROPERTY);
         }
-        if (GREMLIN.equalsIgnoreCase(language) || SQL_GREMLIN.equalsIgnoreCase(language)) {
-            return GremlinConnectionProperties.CONTACT_POINT_KEY;
+
+        private Properties setProperties(final String propertyString, final String firstPropertyKey,
```

Comment:
I could be lacking context here, but is the naming of this correct? It sounds like a setter but is returning the Properties as well.

Reply:
It sets and returns the property, just for the class though. Is there a better convention to follow?

Reply:
It takes properties from string, and makes them into a property to be returned. It's just a helper for the above function. Is there a different naming convention I should use to make it more clear?